formats/cbor/commonMain/src/kotlinx/serialization/cbor/internal/Encoding.kt

```diff
@@ -46,91 +48,316 @@ private const val SINGLE_PRECISION_MAX_EXPONENT = 0xFF
         return true
     }
 
-    override fun encodeString(value: String) = encoder.encodeString(value)
+    private fun encodePreamble(
+        parentDescriptor: SerialDescriptor?,
+        index: Int,
+        label: Long?,
+        name: String?
+    ): ByteArray {
+        val preamble = ByteArrayOutput()
+        val encoder = CborEncoder(preamble)
+
+        parentDescriptor?.let { descriptor ->
+
+
+            if (!descriptor.hasArrayTag()) {
+                if (cbor.writeKeyTags) {
+                    index.let { descriptor.getKeyTags(it)?.forEach { encoder.encodeTag(it) } }
+                }
+                if ((descriptor.kind !is StructureKind.LIST) && (descriptor.kind !is StructureKind.MAP) && (descriptor.kind !is PolymorphicKind)) {
+                    //indices are put into the name field. we don't want to write those, as it would result in double writes
+                    if (cbor.preferSerialLabelsOverNames && label != null) {
+                        encoder.encodeNumber(label)
+                    } else if (name != null) {
+                        encoder.encodeString(name)
+                    }
+                }
+            }
+        }
+        if (cbor.writeValueTags) {
+            index.let { parentDescriptor?.getValueTags(it)?.forEach { encoder.encodeTag(it) } }
+        }
+        return preamble.toByteArray()
+    }
 
-    override fun encodeFloat(value: Float) = encoder.encodeFloat(value)
-    override fun encodeDouble(value: Double) = encoder.encodeDouble(value)
+    //If any of the following functions are called for serializing raw primitives (i.e. something other than a class,
+    // list, map or array, no children exist and the root node needs the data
+    override fun encodeString(value: String) {
+        currentToken.apply {
+            data =
+                ByteArrayOutput().also { CborEncoder(it).encodeString(value) }.toByteArray()
 
-    override fun encodeChar(value: Char) = encoder.encodeNumber(value.code.toLong())
-    override fun encodeByte(value: Byte) = encoder.encodeNumber(value.toLong())
-    override fun encodeShort(value: Short) = encoder.encodeNumber(value.toLong())
-    override fun encodeInt(value: Int) = encoder.encodeNumber(value.toLong())
-    override fun encodeLong(value: Long) = encoder.encodeNumber(value)
+        }
+    }
 
-    override fun encodeBoolean(value: Boolean) = encoder.encodeBoolean(value)
+    override fun encodeFloat(value: Float) {
+        currentToken.apply {
+            data =
+                ByteArrayOutput().also { CborEncoder(it).encodeFloat(value) }.toByteArray()
+        }
+    }
 
-    override fun encodeNull() = encoder.encodeNull()
+    override fun encodeDouble(value: Double) {
+        currentToken.apply {
+            data =
+                ByteArrayOutput().also { CborEncoder(it).encodeDouble(value) }.toByteArray()
+        }
+    }
+
+    override fun encodeChar(value: Char) {
+        currentToken.apply {
```

Comment:
`apply` is definitely overused here, can be just `currentToken.data = ...`. But isn't it a bit too much, creating `ByteArrayOutput` and `CborEncoder` for every write operation? Maybe you can refactor a code a little bit so you can just use `encodeNumber` and get a byte array directly.

Reply:
Refactored in d3039fb